### DISCUSSION TITLE
Subscriptions filter

### DISCUSSION THREAD

**Discussion by gmargriff** [UltimaHoarder/UltimaScraper#1232] [Q&A] [answered]
I know it isn't straight connected to the scope, but I guess it won't hurt to ask. I've been using the subscriptions module to get a list of subscribers from page, it works fine until I try to use a filter. 

When I try to reach 
`https://onlyfans.com/api2/v2/subscriptions/subscribers?limit=100&offset=0&type=all` 

it works just great, but there are still too many pages to check, so I add 
`https://onlyfans.com/api2/v2/subscriptions/subscribers?limit=10&offset=0&type=all&filter[total_spent]=10` 

and it just falls back to "401 Please refresh the page".

Can anyone, please, point me to the right direction here?

Thanks!

**Comment by UltimaHoarder** [comment, marked as answer]
Are you modifying my script or is it your own custom code?

**Reply by gmargriff** [reply]
Thanks for your reply! It's custom, using yours to manage sessions and requests, so I added that to the apis/onlyfans/onlyfans.py

```
def get_subscriptions_list(self, limit: int = 99, offset: int = 0, total_spent: int = 0):
        if not self.active:
            return
        link = endpoint_links(global_limit=limit,
                              global_offset=offset, total_spent=total_spent).subscriptions
        session = self.session_manager.sessions[0]
        results = self.session_manager.json_request(link)
        return results
```

and added custom parameters to subscriptions' endpoint_links 
`self.subscriptions = f"https://onlyfans.com/api2/v2/subscriptions/subscribes?limit={global_limit}&offset={global_offset}&type=active"`

I'm calling it after OF session is created, works fine even with bigger limits (1000 per page), but not with the filter[] parameter :/

**Reply by UltimaHoarder** [reply]
Ah I know why. Some requests require the user-Id header. 
Unfortunately, I took it out.

You'd have to add user-id to authed.session_manager.headers before you send the get_subscriptions function

**Reply by gmargriff** [reply]
Thanks again for your reply! Tried to add user-id header to request, still 401... my headers ended up with something like this...

`{'user-agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36', 'referer': 'https://onlyfans.com/api2/v2/users/me', 'x-bc': 'xbcfromlocalstorage', 'app-token': '33d57ade8c02dbc5a333db99ff9ae26a', 'sign': 'generatedsign', 'time': '1621617794', 'user-id': 'authidfromcookies'}`

Also, had to add header on request, if I remove `headers.pop(remove_header)` from `create_headers()`  it won't authenticate... I'm still missing something... (new to Python, kinda struggling)

`{'error': {'code': 401, 'message': 'Please refresh the page'}}`

**Reply by UltimaHoarder** [reply]
Are you only using the API and porting it to your project or are you customising my script?
And ye you can't auth without removing the id.
You have to auth first and then add back the headers.

**Reply by gmargriff** [reply]
For the first part, only using the API, figured out (kinda) which part does which function then added a new function to apis/onlyfans/onlyfans.py to not break working functions.

My outside main file is something like.:
```
for authed in api.auths:
        data = []
        offset = 0
        counter = 0
        while offset < 1:
            data = authed.get_subscriptions_list(limit = 100, offset = offset)
```

and to the custom apis/onlyfans/onlyfans.py I've added.:
```
def get_subscriptions_list(self, check: bool = False, refresh: bool = True, limit: int = 99, offset: int = 0, total_spent: int = 0):
        if not self.active:
            return
        link = endpoint_links(global_limit=limit,
                              global_offset=offset, total_spent=total_spent).subscriptions
        session = self.session_manager.sessions[0]
        results = self.session_manager.json_request(link)
        print(results)
        return results
```

Just like that it works fine, but when adding filter parameter it breaks... added `user-id` header into json_request function conditionally, so it just adds when sending that specific request (also, to not break working requests) but still same problem...

**Reply by gmargriff** [reply]
Man, I'm so f*kng dumb, I've misread the route, the filter I'm trying to use don't even exist in this route..........
